libbpf-tools/trace_helpers.c

```diff
@@ -169,6 +175,604 @@ const struct ksym *ksyms__get_symbol(const struct ksyms *ksyms,
+{
+	return !strcmp(path, "[vdso]");
+}
+
+static int get_elf_text_scn_info(const char *path, uint64_t *addr,
+				 uint64_t *offset)
+{
+	Elf_Scn *section = NULL;
+	int fd = -1, err = -1;
+	GElf_Shdr header;
+	size_t stridx;
+	Elf *e = NULL;
+	char *name;
+
+	e = open_elf(path, &fd);
+	if (!e)
+		goto err_out;
+	err = elf_getshdrstrndx(e, &stridx);
+	if (err < 0)
+		goto err_out;
+
+	err = -1;
+	while ((section = elf_nextscn(e, section)) != 0) {
+		if (!gelf_getshdr(section, &header))
+			continue;
+
+		name = elf_strptr(e, stridx, header.sh_name);
+		if (name && !strcmp(name, ".text")) {
+			*addr = (uint64_t)header.sh_addr;
+			*offset = (uint64_t)header.sh_offset;
+			err = 0;
+			break;
+		}
+	}
+
+err_out:
+	close_elf(e, fd);
+	return err;
+}
+
+static int syms__add_dso(struct syms *syms, struct map *map, const char *name)
+{
+	struct dso *dso = NULL;
+	int i, type;
+	void *tmp;
+
+	for (i = 0; i < syms->dso_sz; i++) {
+		if (!strcmp(syms->dsos[i].name, name)) {
+			dso = &syms->dsos[i];
+			break;
+		}
+	}
+
+	if (!dso) {
+		tmp = realloc(syms->dsos, (syms->dso_sz + 1) *
+			      sizeof(*syms->dsos));
+		if (!tmp)
+			return -1;
+		syms->dsos = tmp;
+		dso = &syms->dsos[syms->dso_sz++];
+		memset(dso, 0, sizeof(*dso));
+		dso->name = strdup(name);
+	}
+
+	tmp = realloc(dso->ranges, (dso->range_sz + 1) * sizeof(*dso->ranges));
+	if (!tmp)
+		return -1;
+	dso->ranges = tmp;
+	dso->ranges[dso->range_sz].start = map->start_addr;
+	dso->ranges[dso->range_sz].end = map->end_addr;
+	dso->ranges[dso->range_sz].file_off = map->file_off;
+	dso->range_sz++;
+	type = get_elf_type(name);
+	if (type == ET_EXEC) {
+		dso->type = EXEC;
+	} else if (type == ET_DYN) {
+		dso->type = DYN;
+		if (get_elf_text_scn_info(name, &dso->sh_addr, &dso->sh_offset) < 0)
+			return -1;
+	} else if (is_perf_map(name)) {
+		dso->type = PERF_MAP;
+	} else if (is_vdso(name)) {
+		dso->type = VDSO;
+	} else {
+		dso->type = UNKNOWN;
+	}
+	return 0;
+}
+
+static struct dso *syms__find_dso(const struct syms *syms, unsigned long addr,
+				  uint64_t *offset)
+{
+	struct load_range *range;
+	struct dso *dso;
+	int i, j;
+
+	for (i = 0; i < syms->dso_sz; i++) {
+		dso = &syms->dsos[i];
+		for (j = 0; j < dso->range_sz; j++) {
+			range = &dso->ranges[j];
+			if (addr <= range->start || addr >= range->end)
+				continue;
+			if (dso->type == DYN || dso->type == VDSO) {
+				/* Offset within the mmap */
+				*offset = addr - range->start + range->file_off;
+				/* Offset within the ELF for dyn symbol lookup */
+				*offset += dso->sh_addr - dso->sh_offset;
+			} else {
+				*offset = addr;
+			}
+
+			return dso;
+		}
+	}
+
+	return NULL;
+}
+
+static int dso__load_sym_table_from_perf_map(struct dso *dso)
+{
+	return -1;
+}
+
+static int dso__add_sym(struct dso *dso, const char *name, uint64_t start,
+			uint64_t size)
+{
+	size_t new_cap, name_len = strlen(name) + 1;
+	struct sym *sym;
+	void *tmp;
+
+	if (dso->strs_sz + name_len > dso->strs_cap) {
```

Comment:
so, libbpf's `struct btf` is actually a pretty efficient "set of strings" data structure, how about using it instead? see btf__add_str() and btf__find_str().

Reply:
Cool, changed to this, thanks!